optuna/multi_objective/trial.py

```diff
@@ -0,0 +1,218 @@
+from datetime import datetime
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Sequence
+from typing import Union
+
+from optuna._experimental import experimental
+from optuna.distributions import BaseDistribution
+from optuna import multi_objective
+from optuna.structs import FrozenTrial
+from optuna.structs import StudyDirection
+from optuna.structs import TrialState
+from optuna.trial import Trial
+
+CategoricalChoiceType = Union[None, bool, int, float, str]
+
+
+@experimental("1.4.0")
+class MultiObjectiveTrial(object):
+    def __init__(self, trial: Trial):
+        self._trial = trial
+        self._n_objectives = multi_objective.study.MultiObjectiveStudy(trial.study).n_objectives
+
+    def suggest_uniform(self, name: str, low: float, high: float) -> float:
+        return self._trial.suggest_uniform(name, low, high)
+
+    def suggest_loguniform(self, name: str, low: float, high: float) -> float:
+        return self._trial.suggest_loguniform(name, low, high)
+
+    def suggest_discrete_uniform(self, name: str, low: float, high: float, q: float) -> float:
+        return self._trial.suggest_discrete_uniform(name, low, high, q)
+
+    def suggest_int(self, name: str, low: int, high: int) -> int:
+        return self._trial.suggest_int(name, low, high)
+
+    def suggest_categorical(
+        self, name: str, choices: Sequence[CategoricalChoiceType]
+    ) -> CategoricalChoiceType:
+        return self._trial.suggest_categorical(name, choices)
+
+    def report(self, values: List[float], step: int) -> None:
+        if len(values) != self._n_objectives:
+            raise ValueError(
+                "The number of the intermediate values {} at step {} is mismatched with"
+                "the number of the objectives {}.",
+                len(values),
+                step,
+                self._n_objectives,
+            )
+
+        for i, value in enumerate(values):
+            self._trial.report(value, self._n_objectives * (step + 1) + i)
+
+    def _report_complete_values(self, values: List[float]) -> None:
+        if len(values) != self._n_objectives:
+            raise ValueError(
+                "The number of the values {} is mismatched with the number of the objectives {}.",
+                len(values),
+                self._n_objectives,
+            )
+
+        for i, value in enumerate(values):
+            self._trial.report(value, i)
```

Comment:
I'm wondering if the reserving of the first steps to the final value doesn't complicate the implementation of pruners. Maybe I'm missing something but would for instance using `system_attrs` work?

Reply:
I assume that pruners (that are multi-objective version) would access intermediate values via `FrozenMultiObjectiveTrial.intermediate_values` field, not the underlying `FrozenTrial.intermediate_values` (please refer to [the constructor of the class](https://github.com/optuna/optuna/pull/1054/files#diff-2ba765a95ea5e65aa7f34eb52baf57f1R111) for the field detail). 
Pruners would not have to care about the encoding detail because they would not see the raw representation of the intermediate values.

Reply:
Right, it's hidden the from pruners. Thanks.